File: HouseRobber.js
/**
 * @param {number[]} nums
 * @return {number}
 */


// Approach: We will use the dynamic programming approach using array
// Time Complexity is O(n) where n is the length of the nums array
// Space Complexity is O(1) as we are using only constant space
var rob = function(nums) {
    if(nums.length == 1) return nums[0];
    for(let i=0;i<nums.length;i++){
        if(i == 1) nums[i] = Math.max(nums[i],nums[i-1]);
        if(i > 1){
            nums[i] = Math.max(nums[i-1], nums[i]+nums[i-2]);
        }
    }
    return nums[nums.length-1];
};


// Approach: We will use the dynamic programming approach using two variables
// Time Complexity is O(n) where n is the length of the nums array
// Space Complexity is O(1) as we are using only constant space
var rob = function(nums) {
    if(nums.length == 1) return nums[0];
    let prev = 0;
    let curr = 0;
    for(let i=0;i<nums.length;i++){
        let temp = curr;
        curr = Math.max(curr, nums[i]+prev);
        prev = temp;
    }
    return curr;
};

File: houseRobber.cpp
//TC o(n)
//SC o(1)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No
//In this approach we will use two int instead of creating a dp matrix to optimize sc
//Since we have only 1 deciding factor we will go through every option and check for max 
//for choose and no choose case and will return the max in the end
class Solution {
public:
    int rob(vector<int>& nums) {
        int n = nums.size();
        if(n==1)return nums[0];
        vector<int>dp(n);

        int prev2 = nums[0];
        int prev1 = max(nums[0],nums[1]);
        for(int i =2;i<n;i++){
            int curr = max(prev1,prev2+nums[i]);
            prev2 = prev1;
            prev1 = curr;
        }
        return prev1;
    }
};
